Clarify simple function call. (#3897)
modelica/ModelicaSpecification: chapters/equations.tex

Also compare with \cref{assignments-from-called-functions-with-multiple-results} about calling functions with multiple outputs in assignment statements.
\end{nonnormative}

> Note that there is also \lstinline!function_name(inexpr1, inexpr2);! which calls the given function or operator.

Reviewer: I don't think a normative paragraph like this should start with _Note that_. I also don't understand why this form would be presented as a form of simple equality equation.

```suggestion

\subsection{Side-Effect Equations}\label{side-effect-equations}

An equation in the following form calls the given function or operator:
\begin{itemize}
\item \lstinline!function_name(inexpr1, inexpr2);!
\end{itemize}
```

Author: I'm not sure if "Side Effect Equations" is the best name, but I cannot find any better - so let's use it.

It behaves similarly to a simple equation (calling the function or operator), even if it uses a different production in the grammar.
It is used for special operators like \lstinline!reinit! (\cref{reinit}), \lstinline!assert! (\cref{assert}), and \lstinline!terminate! (\cref{terminate}).

> \begin{nonnormative}

Reviewer: Empty line for source code readability:

```suggestion

\begin{nonnormative}
```

That case is not recommended, but it is allowed to avoid having special rules in function-subtyping \cref{function-compatibility-or-function-subtyping-for-functions}.
\end{nonnormative}

> It is used for special operators like \lstinline!reinit! (\cref{reinit}), \lstinline!assert! (\cref{assert}), and \lstinline!terminate! (\cref{terminate}).

Author: ```suggestion
It is used for special operators like \lstinline!reinit! (\cref{reinit}), \lstinline!assert! (\cref{assert}), and \lstinline!terminate! (\cref{terminate}).
\begin{nonnormative}
If the function has any outputs they will be ignored.
That case is not recommended, but it is allowed to avoid having special rules in function-subtyping \cref{function-compatibility-or-function-subtyping-for-functions}.
\end{nonnormative}
```
Reordering the text for two reasons:
- The reinit, assert etc is more important.
- The "It" in "It is used..." was a bit confusing after the non-normative case.

Also compare with \cref{assignments-from-called-functions-with-multiple-results} about calling functions with multiple outputs in assignment statements.
\end{nonnormative}

Note that there is also \lstinline!function_name(inexpr1, inexpr2);! which calls the given function or operator.

> It behaves similarly to a simple equation (calling the function or operator but ignoring any potential output), even if it uses a different production in the grammar.

Reviewer: Is this new semantics, or is this already established in some other form somewhere else?  (I am surprised that this would be allowed for a function with outputs; I would have expected that a simple equation/assignment with an empty parenthesis in the left-hand side was needed.)

Author: ```suggestion
It behaves similarly to a simple equation (calling the function or operator), even if it uses a different production in the grammar.
```
I'll have to investigate more to see if it exists, but for now it seems simplest to just remove that part; to not draw attention to it. In general procedure calls have not been handled well.

Note: I assume you specifically mean the part about ignoring outputs; as the other semantics is already used in MSL, e.g., for `Modelica.Utilities.Streams.print`, `Modelica.Utilities.Streams.error` (and some assert-media function as far as I recall).

Ignoring outputs is _implicitly_ assumed in https://specification.modelica.org/master/interface-or-type-relationships.html#function-compatibility-or-function-subtyping-for-functions - because otherwise a function adding a new output after the last one would need a sub-typing exception for functions with no outputs.

Author: Since it is delayed we can discuss and clarify it further, including stating that procedure call for functions with outputs is not recommended, but simplifies function sub-typing.

Author: The extra idea would be:
```suggestion
It behaves similarly to a simple equation (calling the function or operator), even if it uses a different production in the grammar.
\begin{nonnormative}
If the function has any outputs they will be ignored.
That case is not recommended, but it is allowed to avoid having special rules in function-subtyping \cref{function-compatibility-or-function-subtyping-for-functions}.
\end{nonnormative}
```

Author: Should be ok now.

> It is used for special operators like \lstinline!reinit! (\cref{reinit}), \lstinline!assert! (\cref{assert}), and \lstinline!terminate! (\cref{terminate}).

Reviewer: Restore empty line:

```suggestion
It is used for special operators like \lstinline!reinit! (\cref{reinit}), \lstinline!assert! (\cref{assert}), and \lstinline!terminate! (\cref{terminate}).

```
